clash-cores/src/Clash/Cores/UART.hs

```diff
@@ -0,0 +1,205 @@
+                             }
+      | otherwise        = state
+                             { uartTxPeriodCounter = uartTxPeriodCounter + 1 }
+
+-- | State transition function for the UART transmitter
+uartTx
+  :: forall (baudRate :: Nat)
+            (dom :: Domain)
+   . ( HiddenClockResetEnable dom
+     , KnownNat baudRate
+     , KnownNat (DomainPeriod dom)
+     , 1 <= baudRate
+     , 1 <= (DomainPeriod dom)
+     )
+  => Signal dom (Maybe (BitVector 8))
+  -- ^ The word to transmit
+  -> (Signal dom Bit, Signal dom Bool)
+  -- ^ The output bits for the tx line and an acknowledgement that the word
+  -- to transmit has been received
+uartTx mWord = let uartTx' = uartTxT @(Div (HzToPeriod baudRate) (DomainPeriod dom))
+               in unbundle $ register (1, False) $ mealy uartTx' UartTxIdle mWord
+
+-- | State used for the UART receiver
+data UartRxState (period :: Nat)
+   -- | Waiting for a start bit
+   = UartRxIdle
+   -- | Wait until we are at the center of the start bit
+   | UartRxStartBit
+       { uartRxHalfPeriodCounter :: Index (Div period 2)
+       -- ^ Counter that overflows at half a UART period
+       }
+   -- | Receive data bits
+   | UartRxDataBits
+       { uartRxWord :: BitVector 8
+       -- ^ The data received
+       , uartRxPeriodCounter :: Index period
+       -- ^ Counter that overflows after a full uart period
+       , uartRxWordCounter :: Index 9
+       -- ^ Which bit of uart we are currently waiting for. This is 9 because
+       -- we also wait until we check the stop bit.
+       }
+   -- | Check the stop bit is correct
+   | UartRxStopBit
+       { uartRxWord :: BitVector 8
+       -- ^ The received word
+       }
+  deriving (Show, Generic, NFDataX)
+
+-- | State transition function for the UART receiver
+uartRxT
+  :: forall (period :: Nat)
+   . KnownNat period
+  => UartRxState period
+  -- ^ The current state
+  -> Bit
+  -- ^ The uart receive line
+  -> (UartRxState period, Maybe (BitVector 8))
+  -- ^ The next state and a potential word that has been received
+uartRxT UartRxIdle 1 = (UartRxIdle, Nothing)
+uartRxT UartRxIdle 0 = (UartRxStartBit 1, Nothing)
+uartRxT UartRxStartBit{..} b = (nextState, Nothing)
+  where
+    halfPeriodDone = uartRxHalfPeriodCounter == maxBound
+    stillLow = b == 0
+    validStartBit = halfPeriodDone && stillLow
+
+    nextState | validStartBit  = UartRxDataBits
+                                   { uartRxWord = 0
+                                   , uartRxPeriodCounter = 0
+                                   , uartRxWordCounter = 0
+                                   }
+              | halfPeriodDone = UartRxIdle
+              | otherwise      = UartRxStartBit
+                                   { uartRxHalfPeriodCounter = uartRxHalfPeriodCounter + 1 }
+uartRxT state@UartRxDataBits{..} b = (nextState, Nothing)
+  where
+    periodDone = uartRxPeriodCounter == maxBound
+    lastBit = uartRxWordCounter == maxBound
```

Comment:
Getting rid of `uartRxWordCounter` is a bit more involved than getting rid of `uartTxWordCounter`, but still possible. If we initialize `uartRxWord` to `1 :> repeat 0`, we can tell that we've gotten all databits as soon as we shift out a 1; while we are still receiving a word, we shift out zeroes.